File: src/crypto/wordcoding/WordCode.ts
import { dictName as englishDictName,
         words as englishWords } from './dicts/english';
import { dictName as spanishDictName, 
         normalizer as spanishNormalizer, 
         words as spanishWords } from './dicts/spanish';



class WordCode {

    static english = new WordCode(englishDictName, englishWords);
    static spanish = new WordCode(spanishDictName, spanishWords, spanishNormalizer);

    static lang = new Map<String, WordCode>([['es', WordCode.spanish], ['en', WordCode.english]]);
    
    static all  = [WordCode.english, WordCode.spanish];

    dictName : string;
    words    : string[];

    wordPositions?: Map<string, number>;

    bitsPerWord: number;
    normalizer: (word: string) => string;

    constructor(dictName: string, words: string[], normalizer?: (word: string) => string) {
        this.dictName = dictName;
        this.words    = words;
        this.bitsPerWord = Math.log2(this.words.length);
        this.normalizer = normalizer === undefined? (x: string) => x.toLowerCase() : normalizer;
    }

    private fillWordPositions(): void {
        if (this.wordPositions === undefined) {

            this.wordPositions = new Map<string, number>();
            let pos=0;
            for (const word of this.words) {
                this.wordPositions.set(this.normalizer(word), pos);
                pos = pos + 1;
            }
        }
    } 

    // encode: get a hex string (containing a multiple of bitsPerWord bits), 
    // and get a sequence of words encoding it
    encode(hex: string): string[] {

        const nibblesPerWord = this.bitsPerWord / 4;

        let wordNibbles = '';
        let words: string[] = [];

        for (let i=0 ; i<hex.length; i++) {
            wordNibbles = wordNibbles + hex[i];
            if (wordNibbles.length === nibblesPerWord) {
                words.push(this.encodeWord(wordNibbles));
                wordNibbles='';
            }
        }

        if (wordNibbles.length !== 0) {
            throw new Error('Trying to word-encode a hex string whose lenght does not correspond to a multiple of the bits-per-word constant.');
        }

        return words;
    }

    private encodeWord(hex: string): string {

        const pos = Number.parseInt(hex, 16);

        if (pos >= this.words.length) {
            throw new Error('Number is too large to encode as a single word');
        }

        return this.words[pos];
    }

    // decode: get a sequence of words, return the hex value they encode.
    decode(words: string[]): string {
        this.fillWordPositions();

        let result = '';

        const nibblesPerWord = this.bitsPerWord / 4;

        for (let word of words) {
            let position = this.wordPositions?.get(this.normalizer(word));

            if (position === undefined) {
                throw new Error('Trying to decode wordcoded number but received a word that is not in the dictionary "' + this.dictName + '":' + word);
            }

            result = result + position.toString(16).padStart(nibblesPerWord, '0');
        }

        return result.toUpperCase();
    }
    
    check(word: string) {
        this.fillWordPositions();

        return this.wordPositions?.get(this.normalizer(word)) !== undefined;
    }
}

export { WordCode };

File: src/crypto/wordcoding/WordCode.rs
use std::collections::HashMap;

pub struct WordCode {
    dict_name: String,
    words: Vec<String>,
    bits_per_word: f64,
    normalizer: Box<dyn Fn(String) -> String>,
    word_positions: Option<HashMap<String, usize>>,
}

impl WordCode {
    pub fn new(
        dict_name: String,
        words: Vec<String>,
        normalizer: Option<Box<dyn Fn(String) -> String>>,
    ) -> Self {
        let bits_per_word = (words.len() as f64).log2();
        let normalizer = normalizer.unwrap_or_else(|| Box::new(|x: String| x.to_lowercase()));

        WordCode {
            dict_name,
            words,
            bits_per_word,
            normalizer,
            word_positions: None,
        }
    }

    fn fill_word_positions(&mut self) {
        if self.word_positions.is_none() {
            let mut word_positions = HashMap::new();
            for (pos, word) in self.words.iter().enumerate() {
                word_positions.insert((self.normalizer)(word.clone()), pos);
            }
            self.word_positions = Some(word_positions);
        }
    }

    pub fn encode(&self, hex: &str) -> Result<Vec<String>, String> {
        let nibbles_per_word = (self.bits_per_word / 4.0).ceil() as usize;

        if hex.len() % nibbles_per_word != 0 {
            return Err("Hex string length is not a multiple of the bits-per-word constant.".to_string());
        }

        let words = hex
            .as_bytes()
            .chunks(nibbles_per_word)
            .map(|chunk| {
                let chunk_str = String::from_utf8_lossy(chunk);
                let pos = usize::from_str_radix(&chunk_str, 16).unwrap();
                self.words.get(pos).unwrap().clone()
            })
            .collect();

        Ok(words)
    }

    pub fn decode(&mut self, words: &[String]) -> Result<String, String> {
        self.fill_word_positions();

        let nibbles_per_word = (self.bits_per_word / 4.0).ceil() as usize;

        let mut result = String::new();

        for word in words {
            let position = self
                .word_positions
                .as_ref()
                .unwrap()
                .get(&(self.normalizer)(word.clone()));

            match position {
                Some(pos) => {
                    result.push_str(&format!("{:0width$X}", pos, width = nibbles_per_word));
                }
                None => {
                    return Err(format!(
                        "Received a word that is not in the dictionary '{}': {}",
                        self.dict_name, word
                    ));
                }
            }
        }

        Ok(result)
    }

    pub fn check(&mut self, word: &str) -> bool {
        self.fill_word_positions();
        self.word_positions.as_ref().unwrap().contains_key(&(self.normalizer)(word.to_string()))
    }
}
